lib/uvm_agents/uvma_rvfi/uvma_rvfi_instr_if.sv

```
@@ -310,6 +328,48 @@ interface uvma_rvfi_instr_if_t
     instr_mem_wmask     <= instr_mem_wmask_f();
   end
 
+  always_comb  cycle_cnt = !reset_n ? 0 : (cycle_cnt_q + 1'b 1);
+
+  always_comb  is_load_instr      = rvfi_valid && |instr_mem_rmask;
+  always_comb  is_store_instr     = rvfi_valid && |instr_mem_wmask;
+  always_comb  is_loadstore_instr = is_load_instr || is_store_instr;
+
+  always_comb  is_mem_act_actual =
+    is_mem_act;  // original signal is already "actual"
+  always_comb  is_mem_act_intended =
+    rvfi_valid && (|instr_mem_rmask || |instr_mem_wmask);
+
+  always_comb  rvfi_pc_upperrdata =
+    (rvfi_insn[1:0] == 2'b 11) ? (
+      rvfi_pc_rdata + 3
+    ) : (
+      rvfi_pc_rdata + 1
+    );
+    // WARNING: Can't trust rvfi_insn if scrambled data.
+    // TODO:WARNING:silabs-robin  Can it be modelled exact?
+
+  always_comb  is_split_instrtrans =
+    rvfi_valid  &&
+    (rvfi_pc_rdata[31:2] != rvfi_pc_upperrdata[31:2]);
+
+  always_comb  is_exception =
+    rvfi_valid  &&
+    rvfi_trap.trap  &&
+    rvfi_trap.exception;
+
+  always_comb  is_instr_acc_fault_pmp =
+    is_exception  &&
+    (rvfi_trap.exception_cause == EXC_CAUSE_INSTR_ACC_FAULT)  &&
+    (rvfi_trap.cause_type == 'h 1);  // TODO:INFO:silabs-robin  Magic num
+
+  always_comb  is_load_acc_fault =
+    is_exception  &&
+    (rvfi_trap.exception_cause == EXC_CAUSE_LOAD_ACC_FAULT);
+
+  always_comb  is_store_acc_fault =
+    is_exception  &&
+    (rvfi_trap.exception_cause == EXC_CAUSE_STORE_ACC_FAULT);
```

The reason for multiple always_combs instead of one big shared begin/end is that errors could be hidden.

I had a case that should've been a "multiple drivers" issue, but since it was in the same begin/end block it was just legal SystemVerilog code.
With individual always_combs that error became a warning instead of something one would manually understand from debugging CEXes.

---

Since these are not including any functions, could this just be simple assigns instead of always_comb-blocks without begin-end statements?

---

(not a fan of always_comb blocks that do not include begin/end statements)

---

Just to make sure I understand, instead of getting an syntax error, you got an cex? So therefore you changed to multiple statements?

---

I made a copy-paste error.
The compiler didn't complain.

If the compiler had given a warning/error instead of being silent, then I could have realized there was a coding mistake instead of spending time on a CEX that didn't make sense.

---

> Since these are not including any functions, could this just be simple assigns instead of always_comb-blocks without begin-end statements?

I'd prefer to have one default safe way to do combinatorial logic than have to juggle between multiple ways and then risk a function sneaking in unnoticed or some other LRM quirk.

I don't mind doing it in one way or the other, as long as my combinatorial signal gets updated combinatorially at all times, no missing sensitivity list, no weird function semantics, etc.

---

> not a fan of always_comb blocks that do not include begin/end statements

Fixed.